**Notas para contenedor de docker:**

Comando de docker para ejecución de la nota de forma local:

nota: cambiar `dir_montar` por la ruta de directorio que se desea mapear a `/datos` dentro del contenedor de docker.

```
dir_montar=<ruta completa de mi máquina a mi directorio>#aquí colocar la ruta al directorio a montar, por ejemplo: 
#dir_montar=/Users/erick/midirectorio.
```

Ejecutar:

```
$docker run --rm -v $dir_montar:/datos --name jupyterlab_prope_r_kernel_tidyverse -p 8888:8888 -d palmoreck/jupyterlab_prope_r_kernel_tidyverse:2.1.4   

```

Ir a `localhost:8888` y escribir el password para jupyterlab: `qwerty`

Detener el contenedor de docker:

```
docker stop jupyterlab_prope_r_kernel_tidyverse
```


Documentación de la imagen de docker `palmoreck/jupyterlab_prope_r_kernel_tidyverse:2.1.4` en [liga](https://github.com/palmoreck/dockerfiles/tree/master/jupyterlab/prope_r_kernel_tidyverse).

---

Para ejecución de la nota usar:

[docker](https://www.docker.com/) (instalación de forma **local** con [Get docker](https://docs.docker.com/install/)) y ejecutar comandos que están al inicio de la nota de forma **local**. 

O bien dar click en alguno de los botones siguientes:

[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/palmoreck/dockerfiles-for-binder/jupyterlab_prope_r_kernel_tidyerse?urlpath=lab/tree/Propedeutico/R/clases/2_probabilidad/6_teorema_de_Bayes.ipynb) esta opción crea una máquina individual en un servidor de Google, clona el repositorio y permite la ejecución de los notebooks de jupyter.

[![Run on Repl.it](https://repl.it/badge/github/palmoreck/dummy)](https://repl.it/languages/Rlang) esta opción no clona el repositorio, no ejecuta los notebooks de jupyter pero permite ejecución de instrucciones de Python de forma colaborativa con [repl.it](https://repl.it/). Al dar click se crearán nuevos ***repl*** debajo de sus users de ***repl.it***.


### Lo siguiente está basado los libros:

### * B. Rumbos, Pensando Antes de Actuar: Fundamentos de Elección Racional, 2009.

### * G. J. Kerns, Introduction to Probability and Statistics Using R, 2014.

### El libro de G. J. Kerns tiene github: [jkerns/IPSUR](https://github.com/gjkerns/IPSUR)

**Notas:**

* Se utilizará el paquete [prob](https://cran.r-project.org/web/packages/prob/index.html) de R para los experimentos descritos en la nota y aunque con funciones nativas de R se pueden crear los experimentos, se le da preferencia a mostrar cómo en R se tienen paquetes para muchas aplicaciones.

* En algunas líneas no es necesario colocar `print` y sólo se ha realizado para mostrar los resultados de las funciones en un formato similar al de R pues la nota se escribió con jupyterlab y R.

* Cuidado al utilizar las funciones del paquete `prob` para construir espacios de probabilidad grandes como lanzar un dado 9 veces... (tal experimento tiene 10 millones de posibles resultados)

In [1]:
options(repr.plot.width=4, repr.plot.height=4) #esta línea sólo se ejecuta para jupyterlab con R

In [2]:
library(prob)

Loading required package: combinat


Attaching package: ‘combinat’


The following object is masked from ‘package:utils’:

    combn


Loading required package: fAsianOptions

Loading required package: timeDate

Loading required package: timeSeries

Loading required package: fBasics

Loading required package: fOptions


Attaching package: ‘prob’


The following objects are masked from ‘package:base’:

    intersect, setdiff, union




# Teorema de Bayes

Supongamos que $E$ y $E^c$ representan los eventos de tener o no una enfermedad viral y que existe una prueba sanguínea para detectar este virus. Denotamos por $+$ al evento de tener una **prueba positiva** y por $-$ al que sea **negativa**. Las probabilidades $P(E)$ y $P(E^c)$ pueden pensarse como las probabilidades **a priori** o **iniciales** de tener o no la enfermedad respectivamente.

La probabilidad $P(E|+)$ cuantifica la probabilidad de tener la enfermedad dado que la prueba de sangre resultó positiva para el virus, es decir, mide **qué tan confiable es el resultado de la prueba** y se llaman probabilidades a **posteriori** o **finales** pues ya **contamos con información** acerca de la prueba de sangre. Una forma de calcularla es la siguiente:

$$P(E|+) = \frac{P(E \cap +)}{P(+)} = \frac{P(+|E)P(E)}{P(+)} = \frac{P(+|E)P(E)}{P(+ \cap E) + P(+ \cap E^c)} = \frac{P(+|E)P(E)}{P(+|E)P(E) + P(+|E^c)P(E^c)}$$

**Obs:** obsérvese que se ha utilizado el **teorema de probabilidad total** en el denominador para calcular $P(+)$. Ver [2_propiedades_de_la_probabilidad](https://github.com/ITAM-DS/Propedeutico/blob/master/R/clases/2_probabilidad/2_propiedades_de_la_probabilidad.ipynb)

Asimismo, $P(E|+), P(E^c|+), P(E|-)$ y $P(E^c|-)$ son probabilidades a posteriori o finales. Las probabilidades a posteriori son una **revisión de las probabilidades originales (las a priori o iniciales)**, actualizadas con la nueva información del resultado de la prueba de sangre. 

En contraste, la probabilidad $P(+|E)$ es la probabilidad de que la prueba resulte positiva dado que tenemos la enfermedad. Este número mide **qué tan confiable es la prueba de sangre**.

El cálculo previo para $P(E|+)$ se le conoce con el nombre de **teorema o fórmula o regla de Bayes**:

Sean $E, E^c$ eventos complementarios de un espacio muestral y sea $F$ algún evento tal que $P(F)\neq 0$, entonces, la probabilidad a posteriori o final de $E$ dado $F$ está dada por:

$$P(E|F) = \frac{P(F|E)P(E)}{P(F)}$$

o equivalentemente, $$P(E|F) = \frac{P(F|E)P(E)}{P(F|E)P(E) + P(F|E^c)P(E^c)}$$

En general si el espacio muestral es la unión de los eventos $E_1, E_2,\dots , E_n$ que son mutuamente excluyentes (nombrados también ajenos), exhaustivos y $F$ es algún evento con $P(F)\neq 0$, la fórmula anterior se generaliza como sigue:

$$P(E_i|F) = \frac{P(F|E_i)P(E_i)}{P(F)} = \frac{P(F|E_i)P(E_i)}{\displaystyle \sum_{j=1}^nP(F|E_j)P(E_j)}$$

**Comentario:** la mayor aportación de **Thomas Bayes** (1702-1761) radica en la idea de inferencia inversa, que va de los efectos a las causas, en lugar de ir de las causas a los efectos. Si $E$ es la causa y $F$ el efecto, $P (E | F )$ representa la probabilidad de la causa dado que se observó el efecto. El teorema de Bayes contesta preguntas como, dado que el pavimento está mojado, ¿cuál es la probabilidad de que hubiese llovido? Dado que un testigo es confiable, ¿qué tan confiable es su testimonio? Dado que aumentó el desempleo, ¿qué tan probable es que haya una recesión económica?

## Ejemplos

**1)** La prueba para detectar la presencia de ciertos esteroides en el cuerpo de los atletas tiene falsos positivos, si la prueba es positiva sin haber utilizado esteroides y falsos negativos, si la prueba es negativa a pesar de haber utilizado esteroides. Denotemos por $E$ y $E^c$ a los eventos de haber utilizado o no esteroides y por $+$ y $-$ al resultado de una prueba de sangre positiva o negativa, respectivamente. Supongamos que se estima que el 4 % de cierto grupo de atletas ha utilizado esteroides, de manera que: $$P(E)= 0.04, P(E^c) = 0.96$$

Esto nos dice que en un $6\%$ de los casos un atleta que no utiliza esteroides tiene una prueba positiva y, en un $5\%$ un atleta que si los ha utilizado tiene una prueba negativa.

Datos

$P(E) = 0.04$

$P(E^c) = 0.96$

$P(+|E^c) = 0.06$ (falsos positivos)

$P(-|E) = 0.05$ (falsos negativos)

$P(E|+) = \frac{P(+|E)P(E)}{P(+)}$


Supongamos que un atleta es sometido a la prueba y ésta sale positiva.

**a)** ¿cuál es la probabilidad de que haya utilizado esteroides? $P(E|+)$

$P(E|+) = \frac{P(+|E)P(E)}{P(+)}$

$P(E|+) = \frac{P(+|E)P(E)}{P(+|E)P(E) + P(+|E^c)P(E^c)}$

por lo tanto

$P(+|E) = 1 - P(-|E) = 1 - 0.05$

In [9]:
((1 - 0.05)*0.04)/(((1 - 0.05)*0.04) + (0.06*0.96))

[1] 0.3974895

por lo tanto:

$P(E|+) \approx 40\%$

**b)** ¿cuál es la probabilidad que no haya utilizado esteroides? $P(E^c|+)$

$P(E^c|+) = 1 - P(E|+)$

In [10]:
1 - ((1 - 0.05)*0.04)/(((1 - 0.05)*0.04) + (0.06*0.96))

[1] 0.6025105

por lo tanto:

$P(E^c|+) \approx 60\%$

**2)** Considerar el escenario del ejercicio anterior. Suponer que se realiza una segunda prueba, (condicionalmente) independiente de la primera. Calcular $P(E|+ \cap +)$.

$P(E|+ \cap +) \Rightarrow$

$P()$

**3)** Un albino es enjuiciado por haber cometido un crimen. El testigo principal asegura haber visto los hechos a través de una ventana de su casa e identifica a un individuo albino como el criminal. Realizando pruebas para la credibilidad del testigo, se determina que a la distancia y en las condiciones de luz desde las cuales presenció el crimen, identifica correctamente a una persona albina o no albina un 95% de las veces. Aproximadamente una de cada 17 mil personas presenta la condición de albinismo. ¿Cuál es la probabilidad de que el albino haya cometido el crimen dado que el testigo lo identificó?

Datos:

$P(+) \Rightarrow$ probabilidad de que el testigo identifique a un albino

$P(A) = \frac{1}{17,000} \Rightarrow$ probabilidad de que la persona sea albina

$P(+|A) = P(-|A^c) = 95\%$ 

Resultado a obtener:

$P(A|+)$

$P(A|+) = \frac{P(+|A)P(A)}{P(+|A)P(A) + P(+|A^c)P(A^c)}$

$P(+|A^c) = 1 - P(+|A) = 1 - 0.95$

$P(A^c) = 1 - P(A) = 1 - \frac{1}{17,000}$

In [11]:
(0.95*(1/17000))/(0.95*(1/17000) + (1 - 0.95)*(1 - 1/17000))

[1] 0.001116465

Por lo tanto:

$P(A|+) \approx 0.11\%$

## Ejercicios

**Nota: Varios de los siguientes ejercicios tienen su solución en las referencias dadas al inicio de la nota. Se sugiere resolverlos por su cuenta y posteriormente compararlos con sus respectivas soluciones.**

**1)** El siguiente es un mecanismo conocido para elegir a una persona de un grupo pequeño. Digamos que en un grupo de 4 personas quiere seleccionarse a una de ellas para que realice alguna actividad, digamos ir por cervezas a la hielera. Para este efecto, se toman 4 palillos y uno de ellos se recorta; así, se tienen tres palillos largos y uno corto. Las personas toman turnos tomando palillos, ignorando cuál es el corto, y el que saca el palillo corto es el elegido para ir por las cervezas. ¿Tiene alguna ventaja escoger primero, segundo, tercero o cuarto?

- Probabilidad de que el primero saque el palillo corto $\Rightarrow P(p1)$ 

- Probabilidad de que el segundo saque el palillo $\Rightarrow P(p2)$ 

- Probabilidad de que el tercero saque el palillo $\Rightarrow P(p3)$ 

- Probabilidad de que el cuarto saque el palillo $\Rightarrow P(p4)$ 

$P(p1)$ 

In [3]:
1/4

[1] 0.25

$P(p2 \cap p1^c) \Rightarrow P(p2 | p1^c)P(p1^c)$

In [5]:
(1/3)*(3/4)

[1] 0.25

$P(p3 \cap p2^c \cap p1^c) \Rightarrow P(p3 | p2^c \cap p1^c)P(p2^c \cap p1^c) \Rightarrow P(p3 | p1^c \cap p2^c)P(p2^c | p1^c)P(p1^c)$

In [6]:
(1/2)*(2/3)*(3/4)

[1] 0.25

$P(p4 \cap p3^c \cap p2^c \cap p1^c) \Rightarrow $

$P(p4 | p3^c \cap p2^c \cap p1^c)P(p3^c \cap p2^c \cap p1^c) \Rightarrow $

$P(p4 | p3^c \cap p2^c \cap p1^c) P(p3^c|p2^c \cap p1^c)P(p2^c \cap p1^c) \Rightarrow$

$P(p4 | p3^c \cap p2^c \cap p1^c) P(p3^c|p2^c \cap p1^c)P(p2^c | p1^c) P(p1^c) \Rightarrow$

In [7]:
(1)*(1/2)*(2/3)*(3/4)

[1] 0.25

Por lo tanto, no hay ninguna ventaja de perder según el número de palillo que te toque sacar

**2)** Considera el mismo escenario que el ejemplo $3$ pero en lugar de que el sospechoso sea un individuo albino ahora se tiene:

a) El sospechoso del crimen es un individuo pelirrojo y la probabilidad de ser pelirrojo es del $2 \%$. Encontrar $P(\text{\{pelirrojo\}} | +),$ en donde $+$ representa: “el testigo identifica a una persona como pelirroja”.

Datos:

$P(+) \Rightarrow$ probabilidad de que el testigo identifique a un pelirrojo

$P(R) = \frac{2}{100} \Rightarrow$ probabilidad de que la persona sea pelirroja

$P(+|R) = P(-|R^c) = 95\%$ 

Resultado a obtener:

$P(R|+)$

$P(R|+) = \frac{P(+|R)P(R)}{P(+|R)P(R) + P(+|R^c)P(R^c)}$

$P(+|R^c) = 1 - P(+|R) = 1 - 0.95$

$P(R^c) = 1 - P(R) = 1 - \frac{2}{100}$

In [12]:
(0.95*(2/100))/(0.95*(2/100) + (1 - 0.95)*(1 - 2/100))

[1] 0.2794118

Por lo tanto:

$P(R|+) \approx 27.9\%$

b) El sospechoso es un individuo de pelo rubio y la probabilidad de tener el pelo rubio es del $15 \%$. Encontrar $P(\text{\{rubio\}} | +)$, en donde $+$ representa: "el testigo identifica a una persona como rubia".

Datos:

$P(+) \Rightarrow$ probabilidad de que el testigo identifique a un rubio

$P(B) = \frac{15}{100} \Rightarrow$ probabilidad de que la persona sea rubia

$P(+|B) = P(-|B^c) = 95\%$ 

Resultado a obtener:

$P(B|+)$

$P(B|+) = \frac{P(+|B)P(B)}{P(+|B)P(B) + P(+|B^c)P(B^c)}$

$P(+|B^c) = 1 - P(+|B) = 1 - 0.95$ 

$P(B^c) = 1 - P(B) = 1 - \frac{15}{100}$

In [13]:
(0.95*(15/100))/(0.95*(15/100) + (1 - 0.95)*(1 - 15/100))

[1] 0.7702703

Por lo tanto:

$P(B|+) \approx 77\%$